### ISSUE TITLE
`ValueError: node array from the pickle has an incompatible dtype` with sklearn 1.3.0

### ISSUE THREAD

**Issue by maybe-hello-world** [asjacobs92/scikit-explain#3]
After updating sklearn to 1.3.0, I started receiving the following exception:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[17], line 2
      1 trustee = ClassificationTrustee(expert=clf)
----> 2 trustee.fit(x_train, y_train, num_iter=5, num_stability_iter=3, samples_size=0.5)

File /usr/local/lib/python3.10/dist-packages/trustee/main.py:291, in Trustee.fit(self, X, y, top_k, max_leaf_nodes, max_depth, ccp_alpha, train_size, num_iter, num_stability_iter, num_samples, samples_size, use_features, predict_method_name, optimization, aggregate, verbose)
    288     self._top_students.append(max(self._students_by_iter[i], key=lambda item: item[1]))
    290 # Get best overall student based on mean agreement
--> 291 self._best_student = self.explain(top_k=top_k)[0]

File /usr/local/lib/python3.10/dist-packages/trustee/main.py:40, in _check_if_trained.<locals>.wrapper(self, *args, **kwargs)
     38 if len(self._top_students) == 0:
     39     raise ValueError("No student models have been trained yet. Please fit() Trustee explainer first.")
---> 40 return func(self, *args, **kwargs)

File /usr/local/lib/python3.10/dist-packages/trustee/main.py:315, in Trustee.explain(self, top_k)
    293 @_check_if_trained
    294 def explain(self, top_k=10):
    295     """
    296     Returns explainable model that best imitates Expert model, based on highest mean agreement and highest fidelity.
    297 
   (...)
    313             Fidelity of student model to the expert model.
    314     """
--> 315     stable = self.get_stable(top_k=top_k, threshold=0, sort=False)
    316     return max(stable, key=lambda item: item[2])

File /usr/local/lib/python3.10/dist-packages/trustee/main.py:40, in _check_if_trained.<locals>.wrapper(self, *args, **kwargs)
     38 if len(self._top_students) == 0:
     39     raise ValueError("No student models have been trained yet. Please fit() Trustee explainer first.")
---> 40 return func(self, *args, **kwargs)

File /usr/local/lib/python3.10/dist-packages/trustee/main.py:359, in Trustee.get_stable(self, top_k, threshold, sort)
    357 agreement.append([])
    358 # Apply top-k pruning before calculating agreement
--> 359 base_tree = top_k_prune(self._top_students[i][0], top_k=top_k)
    360 for j, _ in enumerate(self._top_students):
    361     # Apply top-k pruning before calculating agreement
    362     iter_tree = top_k_prune(self._top_students[j][0], top_k=top_k)

File /usr/local/lib/python3.10/dist-packages/trustee/utils/tree.py:203, in top_k_prune(dt, top_k, max_impurity)
    200         prune_index(prunned_dt, node["idx"], 0)
    202 # update classifier with prunned model
--> 203 prunned_dt.tree_.__setstate__(get_dt_dict(prunned_dt))
    205 return prunned_dt

File sklearn/tree/_tree.pyx:714, in sklearn.tree._tree.Tree.__setstate__()

File sklearn/tree/_tree.pyx:1418, in sklearn.tree._tree._check_node_ndarray()

ValueError: node array from the pickle has an incompatible dtype:
- expected: {'names': ['left_child', 'right_child', 'feature', 'threshold', 'impurity', 'n_node_samples', 'weighted_n_node_samples', 'missing_go_to_left'], 'formats': ['<i8', '<i8', '<i8', '<f8', '<f8', '<i8', '<f8', 'u1'], 'offsets': [0, 8, 16, 24, 32, 40, 48, 56], 'itemsize': 64}
- got     : [('left_child', '<i8'), ('right_child', '<i8'), ('feature', '<i8'), ('threshold', '<f8'), ('impurity', '<f8'), ('n_node_samples', '<i8'), ('weighted_n_node_samples', '<f8')]
```

Trustee is still correctly working with sklearn 1.2.2. Most likely it's related to changes in serialization in sklearn, see here: https://github.com/scikit-learn/scikit-learn/issues/26798

**Comment by notij** [comment]
@maybe-hello-world If `prunned_dt.tree_.__setstate__(get_dt_dict(prunned_dt))` is self-updating, how about remove this line?

**Comment by maybe-hello-world** [comment]
> @maybe-hello-world If `prunned_dt.tree_.__setstate__(get_dt_dict(prunned_dt))` is self-updating, how about remove this line?

It's more correct to modify `get_dt_dict`, especially this particular line: 
https://github.com/TrusteeML/trustee/blob/a0a17bdd9038608711f8a0bfcc36d6812aaa311f/trustee/utils/tree.py#L109-L114

I'll take a look in near time.

**Comment by notij** [comment]
> > @maybe-hello-world If `prunned_dt.tree_.__setstate__(get_dt_dict(prunned_dt))` is self-updating, how about remove this line?
> 
> It's more correct to modify `get_dt_dict`, especially this particular line:
> 
> https://github.com/TrusteeML/trustee/blob/a0a17bdd9038608711f8a0bfcc36d6812aaa311f/trustee/utils/tree.py#L109-L114
> 
> I'll take a look in near time.

Yes, you are right. But I'm confused about what `get_dt_dict` and `__setstate__` does. I think it might be traversing the tree, extracting the information and then storing them to the tree itself. I tried to compare the tree by exporting text for many test cases, and the result indicates they are the same tree. So I'm wondering if this step is redundant.
<img width="543" alt="image" src="https://github.com/TrusteeML/trustee/assets/90218462/a1938fe5-429f-4c37-9ce0-2843b942272b">
<img width="592" alt="image" src="https://github.com/TrusteeML/trustee/assets/90218462/1fbb6499-ebbd-466e-8bfb-3b5d4d27c264">

**Comment by maybe-hello-world** [comment]
@asjacobs92 i guess we'd appreciate your feedback here :)

**Comment by asjacobs92** [comment]
Hey people, thanks for the comments, and sorry for the delay in responding!

So regarding this:

> But I'm confused about what get_dt_dict and __setstate__ does. I think it might be traversing the tree, extracting the information and then storing them to the tree itself. I tried to compare the tree by exporting text for many test cases, and the result indicates they are the same tree. So I'm wondering if this step is redundant.

It's been a while since I coded this, so take my answer with a grain of salt. But I am pretty sure we still need to set the state.
I think the tree comparison is always true because `tree.export_text` is doing a bit of the work there for you.

What happens is this:
* When I call `prune_index()` to remove the **pruned** branches, I just mark the node as TREE_UNDEFINED.
* This means that the nodes are still in the tree itself, but when you traverse it using recursion, the recursion stops when it finds a TREE_UNDEFINED.
    * When you are calling `tree.export_text()`, it traverses the tree from root to leaves in a recursion, so it won't find any issues.
    * EDIT: link for `export_tree` code. https://github.com/scikit-learn/scikit-learn/blob/f86f41d80bff882689fc16bd7da1fef4a805b464/sklearn/tree/_export.py#L951
* However, if you simply take the node list of the tree, you'll see that the pruned nodes are still in there.
    * This was causing issues for me when I was developing something from the trust report, I just don't remember what exactly.
* So, when I call `get_dt_dict`, I traverse the tree using a recursion to get the correct list of pruned nodes and branches and then use the `__setstate__`  to update the classifier.

So yeah, we still need the __setstate__, and I think we would need to update the dict values. 

I am unsure however how to make this update without breaking it for previous versions of scikit_learn. 
I'll have to take a look into it.

@Xianshun-Jiang for now, if you downgrade the version of scikit-learn, it should unblock you, I think.

Thanks!

**Comment by asjacobs92** [comment]
Issue fixed in version `1.1.5`.
